clang/lib/AST/PreorderAST.cpp

```diff
@@ -256,8 +256,12 @@ void BinaryOperatorNode::Coalesce(bool &Changed, bool &Error) {
     return;
 
   // Coalesce the children first.
-  for (auto *Child : Children)
+  // Since Children is modified within the loop, we need to evaluate
+  // the loop end on each iteration.
+  for (size_t I = 0; I != Children.size(); ++I) {
+    auto *Child = Children[I];
```

Comment:
There is still an incorrectness here - Consider this scenario:
In the third iteration of the loop, `I = 2`. `Child` will be `Children[2]`. Let the `coalesce` function called on `Child` (which is now the same as `Children[2]`) erase `Child`. Because of this erasure, `Children[2]` will now contain what was previously `Children[3]`. Then the control comes out of the `Coalesce`  function and `I` gets incremented to 3. In the next iteration of the loop, `Children[3]` will be processed while `Children[2]` never got processed.

Reply:
Discussed offline. Summary:

Not every child of a `BinaryOperatorNode` `B` will be processed in a call to `B->Coalesce()`. Any children of `B` that were not processed during a call to `B->Coalesce()` will be processed during future calls to `B->Coalesce()` in the `PreorderAST::Normalize` method, which calls `Root->Coalesce()`, `Root->Sort()`, and `Root->ConstantFold()` in a `while` loop while a `Changed` flag is true.

In the future, we may want to rewrite the `for (size_t I = 0; I != Children.size(); ++I)` loop on line 259 so that all children of `B` are processed in a single call to `B->Coalesce()`. This would be done in separate PR(s).